win32/Lib/pywin32_bootstrap.py

```diff
@@ -8,14 +8,20 @@
 # Otherwise, we add this path to PATH.
 import os
 import site
-import sys
 
 # The directory should be installed under site-packages.
-for maybe in site.getsitepackages():
-	pywin32_system32=os.path.join(maybe,"pywin32_system32")
-	if os.path.isdir(pywin32_system32):
-		if hasattr(os, "add_dll_directory"):
-			os.add_dll_directory(pywin32_system32)
-		else:
-			os.environ["PATH"] = pywin32_system32 + ";" + os.environ["PATH"]
 
+dirname = os.path.dirname
+level3_up_dir = dirname(dirname(dirname(__file__)))
```

Comment:
It would be great to get some comments here about why this is done. It also makes sense to only calc the level3 dir if there's no site-packages. Another option I'm 1/2 considering is moving back to just doing the .pth file like in https://github.com/mhammond/pywin32/pull/1430

Reply:
Thanks for the input. Added the comments. When in *Python*'s *venv*, it needs to be calculated as well, because base *Python*'s *site.py* is used, and that doesn't know anything about the (current) virtual environment.

Regarding the alternative you want to go further, it's your call :). Both of them work in the scenarios described by the 2 [SO](https://stackoverflow.com) questions.

Reply:
> When in _Python_'s _venv_, it needs to be calculated as well,

Right - I more meant that the triple dirnames() could be done only after we've determined getsitepackages doesn't exist.

Reply:
I probably wasn't clear enough. In a virtual environment created by [[Python 3.Docs]: venv - Creation of virtual environments](https://docs.python.org/3/library/venv.html#module-venv), `site.getsitepackages` **does** exist, because *site.py* from the *Python* installation that created the current environment is used. But that doesn't "know" anything about the current virtual environment, so its path won't be in the returned list (and as a consequence it has to be added "manually").

```batch
[cfati@CFATI-5510-0:e:\Work\Dev\StackOverflow\q058805040]> "c:\Install\x64\Python\Python\03.07.03\python.exe" -m venv .\stdvenv_py_064_030703

[cfati@CFATI-5510-0:e:\Work\Dev\StackOverflow\q058805040]> .\stdvenv_py_064_030703\Scripts\python.exe -m pip list
Package    Version
---------- -------
pip        19.0.3
setuptools 40.8.0
You are using pip version 19.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

[cfati@CFATI-5510-0:e:\Work\Dev\StackOverflow\q058805040]> .\stdvenv_py_064_030703\Scripts\python.exe -m pip install pywin32
Collecting pywin32
  Downloading https://files.pythonhosted.org/packages/dd/95/690ac56df1b2843a6153de0f22349465175081e39af161465060fe3bea0b/pywin32-226-cp37-cp37m-win_amd64.whl (9.1MB)
    100% |████████████████████████████████| 9.1MB 3.8MB/s
Installing collected packages: pywin32
Successfully installed pywin32-226
You are using pip version 19.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

[cfati@CFATI-5510-0:e:\Work\Dev\StackOverflow\q058805040]> .\stdvenv_py_064_030703\Scripts\python.exe -m pip list
Package    Version
---------- -------
pip        19.0.3
pywin32    226
setuptools 40.8.0
You are using pip version 19.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

[cfati@CFATI-5510-0:e:\Work\Dev\StackOverflow\q058805040]> .\stdvenv_py_064_030703\Scripts\python.exe -c "import win32api"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: DLL load failed: The specified module could not be found.
```

Reply:
Thanks for the clarification. If you have a moment, could you please make the following changes:

* I don't think the `level3_up_dir` in the lambda as the last arg to getattr is strictly necessary - just an empty list would work fine, right, given that the next thing we do is check it?  (This doesn't actually bother me that much, it just jumped out at me)

* I think we should break once the condition at line 26 is true - I meant to add that when I wrote it, and ISTM it becomes slightly more relevant considering venvs. Consider there's a pywin32 installed "globally" (ie, in python itself) *and* pywin32 in the VM. It will never make sense to have both directories on the PATH/dll_directory() - but that also begs the question - which one should come first? It would need to be whichever one is found first for normal module imports, and I assume that's the one in the venv. So that implies to me that we should *insert* the path at the start of `site_packages_dir` rather than appending?

* It would be great if you could also add a quick note to CHANGES.txt and credit yourself in any way you feel comfortable with.

If you have the chance and could quickly test how things work with (say) 225 installed in the "global" python and 226 in the venv, and ensure the "right thing" happens in terms of `pywintypes` and a regular .py file being imported, I'll be happy to merge this and push out a new release.

Thanks for your help!

Reply:
Sorry for the long pause.

Addressed the comments.

Great catch with different versions installed! There were situations when it would have became a problem: a *win32api* version would load another version's *pywintypes*. Not a problem for *v226* / *v225*, but for further apart versions (when *ABI* is no longer compatible) the module would be unusable.

Reply:
@CristiFati That would be because the site-packages aren't yet updated at the point when `pywin32_bootstrap.py` is being executed.

Giving some more details, looking here: https://github.com/python/cpython/blob/master/Lib/site.py#L494-L503
That `addsitepackages(known_paths, [sys.prefix])` invocation will trigger the `.pth` file load/execution down the line, which will in turn call `pywin32_bootstrap.py`.
But `site.getsitepackages()` is based on the `PREFIXES` global variable, which is updated afterwads.

I'd be tempted to consider this a bug in Python/site.py. I was considering opening a bug on the tracker, but I wanted to have a broader look and think it through before sending noise to core python devs :).

Reply:
> Addressed the comments. A note about the _CHANGES.txt_: I bumped the version number to 227. although this change wouldn't quite qualify for a version increase, but I don't know whether it's possible to overwrite files on _PyPI_. Also the 1st record from _build 225_ actually belongs to _226_ (at least that's the 1st version that contains it).

The wording in that could be better - note it is "*since* build XXX" - so your changes should be in "Since build 226" because they will be in since 227 :)  I'll make that change and merge it - thanks!